File: Send-Email-SMS-Azure-AppService-MessageFlow-PHP/MessageFlowSmsEmailApp/services/EmailService.php
<?php

namespace MessageFlow;

use Exception;

/**
 * Email Service for MessageFlow API
 */
class EmailService
{
    private $config;

    public function __construct($config)
    {
        $this->config = $config;

        // Validate configuration
        if (!isset($config['rest_api']['authorization']) || empty($config['rest_api']['authorization'])) {
            throw new Exception("REST API authorization is missing in configuration");
        }
        if (!isset($config['rest_api']['application_key']) || empty($config['rest_api']['application_key'])) {
            throw new Exception("REST API application key is missing in configuration");
        }
    }

    public function sendEmail($request)
    {
        error_log("EmailService.sendEmail called with request: " . json_encode($request, JSON_PRETTY_PRINT));

        try {
            // Prepare headers
            $headers = [
                'Authorization: ' . $this->config['rest_api']['authorization'],
                'Application-Key: ' . $this->config['rest_api']['application_key'],
                'Content-Type: application/json'
            ];

            $authPrefix = substr($this->config['rest_api']['authorization'], 0, 10);
            $appKeyPrefix = substr($this->config['rest_api']['application_key'], 0, 10);
            error_log("Headers set - Authorization: {$authPrefix}..., Application-Key: {$appKeyPrefix}...");

            // Prepare payload
            $payload = [
                'subject' => $request['subject'],
                'smtpAccount' => $request['smtpAccount'],
                'from' => $request['from'],
                'to' => $request['to'],
                'content' => $request['content']
            ];

            // Add optional fields if present
            if (isset($request['tags']) && $request['tags'] !== null) {
                $payload['tags'] = $request['tags'];
            }
            if (isset($request['cc']) && $request['cc'] !== null) {
                $payload['cc'] = $request['cc'];
            }
            if (isset($request['bcc']) && $request['bcc'] !== null) {
                $payload['bcc'] = $request['bcc'];
            }
            if (isset($request['replyTo']) && $request['replyTo'] !== null) {
                $payload['replyTo'] = $request['replyTo'];
            }
            if (isset($request['headers']) && $request['headers'] !== null) {
                $payload['headers'] = $request['headers'];
            }
            if (isset($request['globalVars']) && $request['globalVars'] !== null) {
                $payload['globalVars'] = $request['globalVars'];
            }
            if (isset($request['attachments']) && $request['attachments'] !== null) {
                $payload['attachments'] = $request['attachments'];
            }

            $jsonPayload = json_encode($payload, JSON_PRETTY_PRINT);
            error_log("Email API Request Payload: " . $jsonPayload);

            // Send request using cURL
            error_log("Sending HTTP request to: https://api.messageflow.com/v2.1/email");

            $ch = curl_init('https://api.messageflow.com/v2.1/email');
            curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
            curl_setopt($ch, CURLOPT_POST, true);
            curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($payload));
            curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
            curl_setopt($ch, CURLOPT_TIMEOUT, 30);

            $responseContent = curl_exec($ch);
            $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
            $curlError = curl_error($ch);
            curl_close($ch);

            error_log("Email API Response Status: " . $httpCode);
            error_log("Email API Response Content: " . $responseContent);

            // Handle cURL errors
            if ($curlError) {
                error_log("HTTP request failed: " . $curlError);
                return [
                    'success' => false,
                    'message' => 'HTTP request failed: ' . $curlError,
                    'response_content' => $curlError
                ];
            }

            // Parse response
            if ($httpCode == 200) {
                $responseData = json_decode($responseContent, true);

                if (json_last_error() === JSON_ERROR_NONE && is_array($responseData)) {
                    return [
                        'success' => true,
                        'message' => $responseData['message'] ?? 'Email sent successfully',
                        'response_content' => $responseContent
                    ];
                } else {
                    // If parsing failed but HTTP status is success
                    error_log("Failed to parse response, but HTTP status is success");
                    return [
                        'success' => true,
                        'message' => 'Email sent successfully',
                        'response_content' => $responseContent
                    ];
                }
            } else {
                return [
                    'success' => false,
                    'message' => 'HTTP Error: ' . $httpCode,
                    'response_content' => $responseContent
                ];
            }
        } catch (Exception $ex) {
            error_log("Error sending email: " . $ex->getMessage());
            return [
                'success' => false,
                'message' => 'Error: ' . $ex->getMessage(),
                'response_content' => $ex->getMessage()
            ];
        }
    }
}

File: Send-Email-SMS-Azure-AppService-MessageFlow-Java/MessageFlowSmsEmailApp/src/main/java/com/messageflow/api/services/EmailService.java
package com.messageflow.api.services;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.messageflow.api.models.EmailRequest;
import com.messageflow.api.models.EmailResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.HttpServerErrorException;
import org.springframework.web.client.RestTemplate;

import java.util.HashMap;
import java.util.Map;

@Service
public class EmailService {

    private static final Logger logger = LoggerFactory.getLogger(EmailService.class);
    private static final String EMAIL_API_URL = "https://api.messageflow.com/v2.1/email";

    @Value("${messageflow.api.authorization}")
    private String authorization;

    @Value("${messageflow.api.application-key}")
    private String applicationKey;

    private final RestTemplate restTemplate;
    private final ObjectMapper objectMapper;

    public EmailService(RestTemplate restTemplate, ObjectMapper objectMapper) {
        this.restTemplate = restTemplate;
        this.objectMapper = objectMapper;
    }

    public EmailResponse sendEmail(EmailRequest request) {
        try {
            logger.info("EmailService.sendEmail called with request: {}", objectMapper.writeValueAsString(request));

            // Validate configuration
            if (authorization == null || authorization.isEmpty()) {
                throw new IllegalStateException("REST API authorization is missing in configuration");
            }
            if (applicationKey == null || applicationKey.isEmpty()) {
                throw new IllegalStateException("REST API application key is missing in configuration");
            }

            // Prepare headers
            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_JSON);
            headers.set("Authorization", authorization);
            headers.set("Application-Key", applicationKey);

            logger.info("Headers set - Authorization: {}..., Application-Key: {}...",
                    authorization.substring(0, Math.min(10, authorization.length())),
                    applicationKey.substring(0, Math.min(10, applicationKey.length())));

            // Prepare payload
            Map<String, Object> payload = new HashMap<>();
            payload.put("subject", request.getSubject());
            payload.put("smtpAccount", request.getSmtpAccount());
            payload.put("from", request.getFrom());
            payload.put("to", request.getTo());
            payload.put("content", request.getContent());

            if (request.getTags() != null) {
                payload.put("tags", request.getTags());
            }
            if (request.getCc() != null) {
                payload.put("cc", request.getCc());
            }
            if (request.getBcc() != null) {
                payload.put("bcc", request.getBcc());
            }
            if (request.getReplyTo() != null) {
                payload.put("replyTo", request.getReplyTo());
            }
            if (request.getHeaders() != null) {
                payload.put("headers", request.getHeaders());
            }
            if (request.getGlobalVars() != null) {
                payload.put("globalVars", request.getGlobalVars());
            }
            if (request.getAttachments() != null) {
                payload.put("attachments", request.getAttachments());
            }

            String jsonPayload = objectMapper.writeValueAsString(payload);
            logger.info("Email API Request Payload: {}", jsonPayload);
            logger.info("Sending HTTP request to: {}", EMAIL_API_URL);

            // Send request
            HttpEntity<Map<String, Object>> entity = new HttpEntity<>(payload, headers);
            ResponseEntity<String> response = restTemplate.exchange(
                    EMAIL_API_URL,
                    HttpMethod.POST,
                    entity,
                    String.class
            );

            String responseContent = response.getBody();
            logger.info("Email API Response Status: {}", response.getStatusCode());
            logger.info("Email API Response Content: {}", responseContent);

            // Parse response
            if (response.getStatusCode() == HttpStatus.OK) {
                try {
                    Map<String, Object> responseData = objectMapper.readValue(responseContent, Map.class);
                    String message = responseData.containsKey("message") ?
                            responseData.get("message").toString() : "Email sent successfully";
                    return new EmailResponse(true, message, responseContent);
                } catch (Exception e) {
                    logger.warn("Failed to parse response to EmailResponse model, but HTTP status is success: {}", e.getMessage());
                    return new EmailResponse(true, "Email sent successfully", responseContent);
                }
            } else {
                return new EmailResponse(false, "HTTP Error: " + response.getStatusCode(), responseContent);
            }

        } catch (HttpClientErrorException | HttpServerErrorException ex) {
            logger.error("HTTP request failed: {}", ex.getMessage());
            return new EmailResponse(false, "HTTP Error: " + ex.getStatusCode(), ex.getResponseBodyAsString());
        } catch (Exception ex) {
            logger.error("Error sending email: {}", ex.getMessage());
            return new EmailResponse(false, "Error: " + ex.getMessage(), ex.getMessage());
        }
    }
}
